cmd/duffle/relocate.go

```diff
@@ -0,0 +1,249 @@
+
+	"github.com/deislabs/cnab-go/bundle"
+
+	"github.com/deislabs/duffle/pkg/duffle/home"
+
+	"github.com/spf13/cobra"
+	"k8s.io/apimachinery/pkg/util/validation"
+)
+
+const relocateDesc = `
+Relocates any docker and oci images referenced by a bundle, tags and pushes the images to a registry, and creates a new
+bundle with an updated image map.
+
+The --repository-prefix flag determines the repositories for the relocated images.
+Each image is tagged with a name starting with the given prefix and pushed to the repository.
+
+For example, if the repository-prefix is example.com/user, the image istio/proxyv2 is relocated
+to a name starting with example.com/user/ and pushed to a repository hosted by example.com.
+`
+
+type relocateCmd struct {
+	// args
+	inputBundle  string
+	outputBundle string
+
+	// flags
+	repoPrefix         string
+	inputBundleIsFile  bool
+	outputBundleIsFile bool
+
+	// context
+	home home.Home
+	out  io.Writer
+
+	// dependencies
+	mapping        pathmapping.PathMapping
+	registryClient registry.Client
+}
+
+func newRelocateCmd(w io.Writer) *cobra.Command {
+	relocate := &relocateCmd{out: w}
+
+	cmd := &cobra.Command{
+		Use:   "relocate [INPUT-BUNDLE] [OUTPUT-BUNDLE]",
+		Short: "relocate images in a CNAB bundle",
+		Long:  relocateDesc,
+		Example: `duffle relocate helloworld hellorelocated --repository-prefix example.com/user
+duffle relocate path/to/bundle.json relocatedbundle --repository-prefix example.com/user --input-bundle-is-file
+duffle relocate helloworld path/to/relocatedbundle.json --repository-prefix example.com/user --output-bundle-is-file`,
+		Args: cobra.ExactArgs(2),
+		PreRunE: func(cmd *cobra.Command, args []string) error {
+			// validate --repository-prefix if it is set, otherwise allow flag omission to be diagnosed as such
+			if cmd.Flags().Changed("repository-prefix") {
+				if err := flagValidRepository("repository-prefix")(cmd); err != nil {
+					return err
+				}
+			}
+
+			return nil
+		},
+		RunE: func(cmd *cobra.Command, args []string) error {
+			relocate.inputBundle = args[0]
+			relocate.outputBundle = args[1]
+
+			relocate.home = home.Home(homePath())
+
+			relocate.mapping = pathmapping.FlattenRepoPathPreserveTagDigest
+			relocate.registryClient = registry.NewRegistryClient()
+
+			return relocate.run()
+		},
+	}
+
+	f := cmd.Flags()
+	f.BoolVarP(&relocate.inputBundleIsFile, "input-bundle-is-file", "", false, "Indicates that the input bundle source is a file path")
+	f.BoolVarP(&relocate.outputBundleIsFile, "output-bundle-is-file", "", false, "Indicates that the output bundle destination is a file path")
+	f.StringVarP(&relocate.repoPrefix, "repository-prefix", "r", "", "Prefix for relocated image names")
+	cmd.MarkFlagRequired("repository-prefix")
+
+	return cmd
+}
+
+func (r *relocateCmd) run() error {
+	bun, err := r.setup()
+	if err != nil {
+		return err
+	}
+
+	if err := r.relocate(bun); err != nil {
+		return err
+	}
+
+	if _, err := r.writeBundle(bun); err != nil {
```

Comment:
nit: The first argument (digest) is dropped here and never used, I guess you can then remove it and only return an error.

Reply:
Fixed.